File: backend/app/schemas/master_plan.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class MasterPlanVersionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    version: int
    source: str
    weekly_goals_json: list[dict] | None
    daily_time_budget_json: list[dict] | None
    created_at: datetime


class MasterPlanStateOut(BaseModel):
    plan_id: uuid.UUID | None = None
    plan_status: str | None = None
    active_version: MasterPlanVersionOut | None = None
    pending_version: MasterPlanVersionOut | None = None
    budget_change_ratio: float | None = Field(
        default=None,
        description="Pending vs active total daily minutes change ratio (0-1+).",
    )
    requires_confirmation: bool = False


class MasterPlanConfirmOut(BaseModel):
    active_version: MasterPlanVersionOut
    message: str = "计划已生效"

File: frontend/src/api/masterPlan.ts
import { api } from "./client";

export type MasterPlanVersion = {
  id: string;
  version: number;
  source: string;
  weekly_goals_json: unknown[] | null;
  daily_time_budget_json: { date: string; minutes: number }[] | null;
  created_at: string;
};

export type MasterPlanState = {
  plan_id: string | null;
  plan_status: string | null;
  active_version: MasterPlanVersion | null;
  pending_version: MasterPlanVersion | null;
  budget_change_ratio: number | null;
  requires_confirmation: boolean;
};

export function fetchMasterPlan() {
  return api<MasterPlanState>("/student/master-plan");
}

export function confirmMasterPlan() {
  return api<{ active_version: MasterPlanVersion; message: string }>(
    "/student/master-plan/confirm",
    { method: "POST" },
  );
}

export function rejectMasterPlan() {
  return api<void>("/student/master-plan/reject", { method: "POST" });
}
